File: src/test/java/org/radarcns/hdfs/OffsetRangeSetTest.java
package org.radarcns.hdfs;

import org.junit.Test;
import org.radarcns.hdfs.accounting.OffsetRange;
import org.radarcns.hdfs.accounting.OffsetRangeSet;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class OffsetRangeSetTest {
    @Test
    public void containsGap() {
        OffsetRange am = new OffsetRange("", 0, -1, 0);
        OffsetRange az = new OffsetRange("", 0, 0, 0);
        OffsetRange a = new OffsetRange("", 0, 0, 1);
        OffsetRange c = new OffsetRange("", 0, 3, 4);
        OffsetRange ac = new OffsetRange("", 0, 0, 3);
        OffsetRange ap = new OffsetRange("", 0, 0, 5);
        OffsetRangeSet set = new OffsetRangeSet();
        assertFalse(set.contains(a));
        assertEquals(0, set.size("", 0));
        set.add(a);
        assertEquals(1, set.size("", 0));
        assertTrue(set.contains(a));
        assertTrue(set.contains(az));
        assertFalse(set.contains(c));
        assertFalse(set.contains(ac));
        set.add(c);
        assertEquals(set.toString(), 2, set.size("", 0));
        assertTrue(set.contains(c));
        assertTrue(set.contains(a));
        assertFalse(set.contains(ac));
        set.add(ac);
        assertEquals(1, set.size("", 0));
        set.add(az);
        assertEquals(1, set.size("", 0));
        set.add(am);
        assertEquals(1, set.size("", 0));
        assertTrue(set.contains(ac));
        assertTrue(set.contains(am));
        assertFalse(set.contains(ap));
        set.add(ap);
        assertEquals(1, set.size("", 0));
    }

}

File: src/test/java/org/radarbase/output/OffsetRangeSetTest.kt
package org.radarbase.output

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.radarbase.output.accounting.OffsetRangeSet
import org.radarbase.output.accounting.TopicPartition
import org.radarbase.output.accounting.TopicPartitionOffsetRange
import java.time.Instant

class OffsetRangeSetTest {
    private val lastModified = Instant.now()
    private val topicPartition = TopicPartition("", 0)

    @Test
    fun containsGap() {
        val am =
            TopicPartitionOffsetRange(topicPartition, OffsetRangeSet.Range(-1, 0, lastModified))
        val az = TopicPartitionOffsetRange(topicPartition, OffsetRangeSet.Range(0, 0, lastModified))
        val a = TopicPartitionOffsetRange(topicPartition, OffsetRangeSet.Range(0, 1, lastModified))
        val c = TopicPartitionOffsetRange(topicPartition, OffsetRangeSet.Range(3, 4, lastModified))
        val ac = TopicPartitionOffsetRange(topicPartition, OffsetRangeSet.Range(0, 3, lastModified))
        val ap = TopicPartitionOffsetRange(topicPartition, OffsetRangeSet.Range(0, 5, lastModified))
        val set = OffsetRangeSet()
        assertFalse(set.contains(a))
        assertEquals(0, set.size(topicPartition))
        set.add(a)
        assertEquals(1, set.size(topicPartition))
        assertTrue(set.contains(a))
        assertTrue(set.contains(az))
        assertFalse(set.contains(c))
        assertFalse(set.contains(ac))
        set.add(c)
        assertEquals(2, set.size(topicPartition), set.toString())
        assertTrue(set.contains(c))
        assertTrue(set.contains(a))
        assertFalse(set.contains(ac))
        set.add(ac)
        assertEquals(1, set.size(topicPartition))
        set.add(az)
        assertEquals(1, set.size(topicPartition))
        set.add(am)
        assertEquals(1, set.size(topicPartition))
        assertTrue(set.contains(ac))
        assertTrue(set.contains(am))
        assertFalse(set.contains(ap))
        set.add(ap)
        assertEquals(1, set.size(topicPartition))
    }
}
